experiment/clustersecretbackup/main.go

```diff
@@ -0,0 +1,237 @@
+	// ever needed.
+	return fmt.Sprintf("%s__%s__%s", c.clusterContext, clusterSecret.Namespace, clusterSecret.Name)
+}
+
+// gatherOptions parses the command-line flags.
+func gatherOptions(fs *flag.FlagSet, args ...string) options {
+	var o options
+	fs.StringVar(&o.project, "project", "", "GCP project used for backing up secrets")
+	fs.StringVar(&o.clusterContext, "cluster-context", "", "cluster context name used for backing up secrets, must be full form such as <PROVIDER>_<PROJECT>_<ZONE>_<CLUSTER>")
+	fs.StringSliceVar(&o.namespaces, "namespace", []string{}, "namespace to backup, can be passed in repeatedly")
+	fs.StringToStringVar(&o.secrets, "secret-name", nil, "namespace:name of secrets to be backed up, in the form of --secret-name=<namespace>=<name>")
+	fs.BoolVar(&o.update, "update", false, "Controls whether update existing secret or not, if false then secret will only be created")
+	fs.BoolVar(&o.dryRun, "dryrun", false, "Controls whether this is dry run or not")
+	fs.Parse(args)
+
+	return o
+}
+
+// validateFlags validates the command-line flags.
+func (o *options) validateFlags() error {
+	if len(o.project) == 0 {
+		return errors.New("--project must be provided")
+	}
+	if len(o.clusterContext) == 0 {
+		return errors.New("--cluster-context must be provided")
+	}
+	return nil
+}
+
+func newClient(o options) (*client, error) {
+	secretmanagerClient, err := secretmanager.NewClient(o.project)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating secret manager client: %w", err)
+	}
+
+	kubeClient, err := newKubeClients(o.clusterContext)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating kube client: %w", err)
+	}
+	return &client{
+		kubeClient:          kubeClient,
+		secretmanagerClient: secretmanagerClient,
+		options:             o,
+	}, nil
+}
+
+func newKubeClients(clusterContext string) (ctrlruntimeclient.Client, error) {
+	var loader clientcmd.ClientConfigLoader
+	loader = clientcmd.NewDefaultClientConfigLoadingRules()
+
+	cfg, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		loader, &clientcmd.ConfigOverrides{
+			// Enforcing clusterContext in the full form of cluster context,
+			// instead of short names for kubectl.
+			Context:        api.Context{Cluster: clusterContext},
+			CurrentContext: clusterContext,
+		}).ClientConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed create rest config: %v ------ Did you supply the full form of cluster context name instead of short hand?", err)
+	}
+	return ctrlruntimeclient.New(cfg, ctrlruntimeclient.Options{})
+}
+
+// process merges secret into a new secret for write.
+func (c *client) updateSingleSecret(ctx context.Context, clusterSecret *corev1.Secret) error {
+	secretID := c.gsmSecretName(clusterSecret)
+	// Google secret manager expects pure string instead of map[string][]byte,
+	// so translate *corev1.Secret.Data to map[string]string, this will also be
+	// what kubernetes external secret expects.
```

Comment:
Hmm, its too bad that KES doesn't have an option in the ExternalSecret CR to base64 decode the secret from GSM. 
Are we sure that casting binary data to a string, storing in GSM, and then restoring to a k8s secret via KES doesn't mutate the secret data? I'm concerned that casting the `[]byte` to a string before serialization could lose data or cause serialization problems.

Reply:
yeah unfortunately that's the case. I have tried a yaml string as the val in the secret and it worked

Reply:
It makes sense that string data would behave properly, I'm concerned about binary data. Here is an example of the problem: https://play.golang.org/p/WGdM4xTeTV3

Reply:
I see your concern. I think for binary data we'll probably have to choose not using KES. It doesn't hurt backing it up though

Reply:
Sounds good.
/lgtm
/approve